src/R/Components/Impl/Plots/Implementation/RPlotManager.cs

```diff
@@ -181,16 +189,43 @@ internal class RPlotManager : IRPlotManager {
             return props;
         }
 
-        public async Task<LocatorResult> StartLocatorModeAsync(Guid deviceId, CancellationToken cancellationToken) {
-            await _shell.SwitchToMainThreadAsync(cancellationToken);
+        public Task<LocatorResult> StartLocatorModeAsync(Guid deviceId, CancellationToken cancellationToken) {
+            var device = FindDevice(deviceId);
+            device.LocatorMode = true;
 
-            var visualComponent = GetVisualComponentForDevice(deviceId);
-            if (visualComponent == null) {
-                return default(LocatorResult);
-            }
+            _locatorTcs = new TaskCompletionSource<LocatorResult>();
+            _locatorCancelTokenRegistration = cancellationToken.Register(() => CancelLocatorMode(device));
+
+            _shell.DispatchOnUIThread(() => {
+                var visualComponent = GetVisualComponentForDevice(deviceId);
+                visualComponent?.Container.Show(focus: false, immediate: true);
+            });
+
+            return _locatorTcs.Task;
+        }
+
+        public Task EndLocatorModeAsync(IRPlotDevice device, LocatorResult result) {
```

Comment:
Async/await pattern with default task awaiters is quite expensive. Any task awaiter creates garbage in memory, copies all values for `AsyncLocal`, captures context, etc. It also may create additional problems for debugging if stack trace is broken. One of the best practices for async/await is to use it only if it is really required.